src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker_IVsSolutionLoadEvents.cs

```diff
@@ -51,6 +95,217 @@ int IVsSolutionLoadEvents.OnAfterBackgroundSolutionLoadComplete()
+            }
+            OutputToOutputWindow($"Creating projects - done (took {DateTimeOffset.UtcNow - start})");
+
+            OutputToOutputWindow($"Pushing to workspace - start");
+            start = DateTimeOffset.UtcNow;
+            FinishLoad();
+            OutputToOutputWindow($"Pushing to workspace - done (took {DateTimeOffset.UtcNow - start})");
+        }
+
+        private static ImmutableDictionary<string, string> BuildTargetPathMap(IReadOnlyDictionary<string, DeferredProjectInformation> projectInfos)
+        {
+            var builder = ImmutableDictionary.CreateBuilder<string, string>();
+            foreach (var item in projectInfos)
+            {
+                var targetPath = item.Value.TargetPath;
+                if (targetPath != null)
+                {
+                    if (!builder.ContainsKey(targetPath))
+                    {
+                        builder[item.Value.TargetPath] = item.Key;
+                    }
+                    else
+                    {
+                        Debug.Fail($"Already have a target path of '{item.Value.TargetPath}', with value '{builder[item.Value.TargetPath]}'.");
+                    }
+                }
+            }
+            return builder.ToImmutable();
+        }
+
+        [Conditional("DEBUG")]
+        private void OutputToOutputWindow(string message)
+        {
+            _pane?.OutputString(message + Environment.NewLine);
+        }
+
+        private IWorkspaceProjectContext CreateProjectFromArgumentsAndReferences(
+            IWorkspaceProjectContextFactory workspaceProjectContextFactory,
+            string projectFilename,
+            IReadOnlyDictionary<string, DeferredProjectInformation> allProjectInfos,
+            IReadOnlyDictionary<string, string> targetPathsToProjectPaths)
+        {
+            var languageName = GetLanguageOfProject(projectFilename);
+            if (languageName == null)
+            {
+                return null;
+            }
+
+            DeferredProjectInformation projectInfo;
+            if (!allProjectInfos.TryGetValue(projectFilename, out projectInfo))
+            {
+                return null;
+            }
+
+            var commandLineParser = _workspaceServices.GetLanguageServices(languageName).GetService<ICommandLineParserService>();
+            var projectDirectory = Path.GetDirectoryName(projectFilename);
+            var commandLineArguments = commandLineParser.Parse(
+                projectInfo.CommandLineArguments,
+                projectDirectory,
+                isInteractive: false,
+                sdkDirectory: RuntimeEnvironment.GetRuntimeDirectory());
+
+            // TODO: Should come from sln file?
+            var projectName = Path.GetFileNameWithoutExtension(projectFilename);
+            var projectId = GetOrCreateProjectIdForPath(projectFilename, projectName);
+
+            AbstractProject project;
+            if (_projectMap.TryGetValue(projectId, out project))
```

Comment:
It was to handle both.  The as is removed - I changed the return type to `AbstractProject` to make it clearer.